modules/queue/queue_wrapped.go

```diff
@@ -0,0 +1,196 @@
+			}
+			sleepTime := 100 * time.Millisecond
+			if q.timeout > 0 && q.maxAttempts > 0 {
+				sleepTime = (q.timeout - 200*time.Millisecond) / time.Duration(q.maxAttempts)
+			}
+			time.Sleep(sleepTime)
+		}
+	}
+}
+
+// WrappedQueue wraps a delayed starting queue
+type WrappedQueue struct {
+	delayedStarter
+	handle   HandlerFunc
+	exemplar interface{}
+	channel  chan Data
+}
+
+// NewWrappedQueue will attempt to create a queue of the provided type,
+// but if there is a problem creating this queue it will instead create
+// a WrappedQueue with delayed startup of the queue instead and a
+// channel which will be redirected to the queue
+func NewWrappedQueue(handle HandlerFunc, cfg, exemplar interface{}) (Queue, error) {
+	configInterface, err := toConfig(WrappedQueueConfiguration{}, cfg)
+	if err != nil {
+		return nil, err
+	}
+	config := configInterface.(WrappedQueueConfiguration)
+
+	queue, err := CreateQueue(config.Underlying, handle, config.Config, exemplar)
+	if err == nil {
+		// Just return the queue there is no need to wrap
+		return queue, nil
+	}
+	if IsErrInvalidConfiguration(err) {
+		// Retrying ain't gonna make this any better...
+		return nil, ErrInvalidConfiguration{cfg: cfg}
+	}
+
+	queue = &WrappedQueue{
+		handle:   handle,
+		channel:  make(chan Data, config.QueueLength),
+		exemplar: exemplar,
+		delayedStarter: delayedStarter{
+			cfg:         config.Config,
+			underlying:  config.Underlying,
+			timeout:     config.Timeout,
+			maxAttempts: config.MaxAttempts,
+			name:        config.Name,
+		},
+	}
+	_ = GetManager().Add(queue, WrappedQueueType, config, exemplar, nil)
+	return queue, nil
+}
+
+// Name returns the name of the queue
+func (q *WrappedQueue) Name() string {
+	return q.name + "-wrapper"
+}
+
+// Push will push the data to the internal channel checking it against the exemplar
+func (q *WrappedQueue) Push(data Data) error {
+	if q.exemplar != nil {
+		// Assert data is of same type as r.exemplar
+		value := reflect.ValueOf(data)
+		t := value.Type()
+		exemplarType := reflect.ValueOf(q.exemplar).Type()
+		if !t.AssignableTo(exemplarType) || data == nil {
+			return fmt.Errorf("Unable to assign data: %v to same type as exemplar: %v in %s", data, q.exemplar, q.name)
+		}
+	}
+	q.channel <- data
+	return nil
+}
+
+// Run starts to run the queue and attempts to create the internal queue
+func (q *WrappedQueue) Run(atShutdown, atTerminate func(context.Context, func())) {
+	q.lock.Lock()
+	if q.internal == nil {
+		q.setInternal(atShutdown, q.handle, q.exemplar)
+		go func() {
+			for data := range q.channel {
+				_ = q.internal.Push(data)
+			}
+		}()
+	} else {
+		q.lock.Unlock()
+	}
+
+	q.internal.Run(atShutdown, atTerminate)
+	log.Trace("WrappedQueue: %s Done", q.name)
+}
+
+// Shutdown this queue and stop processing
+func (q *WrappedQueue) Shutdown() {
+	log.Trace("WrappedQueue: %s Shutdown", q.name)
+	q.lock.Lock()
```

Comment:
`Shutdown` can't execute while `setInternal` has a hold on the lock.

Reply:
That's why it adds it's own shutdown handler that causes setInternal to exit